File: packages/plugins/biometrics/Sources/TanoPluginBiometrics/BiometricsPlugin.swift
import Foundation
import TanoBridge

#if canImport(LocalAuthentication)
import LocalAuthentication
#endif

/// Native biometrics plugin for Tano.
///
/// Provides Face ID / Touch ID authentication on iOS devices.
/// On macOS (test environment), returns a success stub.
///
/// Supported methods: `authenticate`.
public final class BiometricsPlugin: TanoPlugin {

    // MARK: - TanoPlugin conformance

    public static let name = "biometrics"
    public static let permissions: [String] = ["biometrics"]

    public init() {}

    // MARK: - Routing

    public func handle(method: String, params: [String: Any]) async throws -> Any? {
        switch method {
        case "authenticate":
            return try await authenticate(params: params)
        default:
            throw BiometricsPluginError.unknownMethod(method)
        }
    }

    // MARK: - authenticate

    private func authenticate(params: [String: Any]) async throws -> [String: Any] {
        guard let reason = params["reason"] as? String else {
            throw BiometricsPluginError.missingParameter("reason")
        }

        #if os(iOS)
        let context = LAContext()
        var error: NSError?

        guard context.canEvaluatePolicy(.deviceOwnerAuthenticationWithBiometrics, error: &error) else {
            throw BiometricsPluginError.biometricsUnavailable(
                error?.localizedDescription ?? "Biometrics not available on this device"
            )
        }

        let success = try await context.evaluatePolicy(
            .deviceOwnerAuthenticationWithBiometrics,
            localizedReason: reason
        )

        return ["success": success]
        #else
        // macOS / test stub — always succeeds
        return ["success": true]
        #endif
    }
}

// MARK: - Errors

public enum BiometricsPluginError: Error, LocalizedError {
    case unknownMethod(String)
    case missingParameter(String)
    case biometricsUnavailable(String)

    public var errorDescription: String? {
        switch self {
        case .unknownMethod(let m):
            return "Unknown biometrics plugin method: \(m)"
        case .missingParameter(let p):
            return "Missing required parameter: \(p)"
        case .biometricsUnavailable(let msg):
            return "Biometrics unavailable: \(msg)"
        }
    }
}

File: packages/android/src/main/kotlin/dev/tano/plugins/BiometricsPlugin.kt
package dev.tano.plugins

import android.content.Context
import dev.tano.bridge.TanoPlugin

/**
 * Native biometrics plugin for Tano on Android.
 *
 * Provides biometric authentication (fingerprint / face). Currently
 * implemented as a stub that always returns success, because real biometric
 * authentication requires a `FragmentActivity` context and UI interaction
 * via `androidx.biometric.BiometricPrompt`.
 *
 * A future version will integrate BiometricPrompt when an Activity reference
 * is available.
 *
 * Supported methods: `authenticate`.
 *
 * Mirrors the iOS BiometricsPlugin.
 *
 * @param context Android Context (FragmentActivity needed for real biometric prompts).
 */
class BiometricsPlugin(private val context: Context) : TanoPlugin {

    override val name: String = "biometrics"
    override val permissions: List<String> = listOf("biometrics")

    // -- Routing --

    override suspend fun handle(method: String, params: Map<String, Any?>): Any? {
        return when (method) {
            "authenticate" -> authenticate(params)
            else -> throw IllegalArgumentException("Unknown biometrics plugin method: $method")
        }
    }

    // -- authenticate --

    /**
     * Authenticates the user via biometrics.
     *
     * Currently a stub that always succeeds. Real implementation requires
     * a FragmentActivity to display the BiometricPrompt dialog.
     */
    private fun authenticate(params: Map<String, Any?>): Map<String, Any?> {
        val reason = params["reason"] as? String
            ?: throw IllegalArgumentException("Missing required parameter: reason")

        // TODO: Integrate BiometricPrompt when Activity reference is available.
        // val executor = ContextCompat.getMainExecutor(context)
        // val prompt = BiometricPrompt(activity, executor, callback)
        // val promptInfo = BiometricPrompt.PromptInfo.Builder()
        //     .setTitle("Authentication")
        //     .setSubtitle(reason)
        //     .setNegativeButtonText("Cancel")
        //     .build()
        // prompt.authenticate(promptInfo)

        // Stub: always return success
        return mapOf("success" to true)
    }
}
